gke-whereami/app.py

```diff
@@ -0,0 +1,146 @@
+                         'instance/hostname',
+                         headers=METADATA_HEADERS)
+        if r.ok:
+            payload['node_name'] = str(r.text)
+    except:
+
+        logging.warning("Unable to capture node name.")
+
+    # get GKE cluster name
+    try:
+        r = requests.get(METADATA_URL +
+                         'instance/attributes/cluster-name',
+                         headers=METADATA_HEADERS)
+        if r.ok:
+            payload['cluster_name'] = str(r.text)
+    except:
+
+        logging.warning("Unable to capture GKE cluster name.")
+
+    # get host header
+    try:
+        payload['host_header'] = request.headers.get('host')
+    except:
+        logging.warning("Unable to capture host header.")
+
+    # get pod name, emoji & datetime
+    payload['pod_name'] = socket.gethostname()
+    payload['pod_name_emoji'] = emoji_list[hash(socket.gethostname()) %
+                                           len(emoji_list)]
+    payload['timestamp'] = datetime.now().replace(microsecond=0).isoformat()
+
+    # get namespace, pod ip, and pod service account via downstream API
+    if os.getenv('POD_NAMESPACE'):
+        payload['pod_namespace'] = os.getenv('POD_NAMESPACE')
+    else:
+        logging.warning("Unable to capture pod namespace.")
+
+    if os.getenv('POD_IP'):
+        payload['pod_ip'] = os.getenv('POD_IP')
+    else:
+        logging.warning("Unable to capture pod IP address.")
+
+    if os.getenv('POD_SERVICE_ACCOUNT'):
+        payload['pod_service_account'] = os.getenv('POD_SERVICE_ACCOUNT')
+    else:
+        logging.warning("Unable to capture pod KSA.")
+
+    # get the whereami METADATA envvar
+    metadata = os.getenv('METADATA')
+    if os.getenv('METADATA'):
+        payload['metadata'] = os.getenv('METADATA')
+    else:
+        logging.warning("Unable to capture metadata.")
+
+    # should we call a backend service?
+    call_backend = os.getenv('BACKEND_ENABLED')
+
+    if call_backend == 'True':
+
+        backend_service = os.getenv('BACKEND_SERVICE')
+
+        try:
+            r = requests.get('http://' + backend_service)
+            if r.ok:
+                backend_result = r.json()
+            else:
+                backend_result = None
+        except:
+
+            print(sys.exc_info()[0])
+            backend_result = None
+
+        payload['backend_result'] = backend_result
+
+    return jsonify(payload)
+
+
+@app.route('/healthz')  # healthcheck endpoint
+def i_am_healthy():
+    return ('OK')
+
+
+if __name__ == '__main__':
+    out_hdlr = logging.StreamHandler(sys.stdout)
+    fmt = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+    out_hdlr.setFormatter(fmt)
+    out_hdlr.setLevel(logging.INFO)
+    logging.getLogger().addHandler(out_hdlr)
+    logging.getLogger().setLevel(logging.INFO)
+    app.logger.handlers = []
+    app.logger.propagate = True
+    app.run(host='0.0.0.0', port=int(os.environ.get('PORT', 8080)))
```

Comment:
Just an idea ...

Would it be better to run and expose this app on port 80? 

I just tried the app out with the one liner and here is what I did:

```
$ kubectl run --image=gcr.io/alexmattson-scratch/whereami --labels app=whereami --expose --port 8080 foo-whereami

# Tested out curling from another pod inside the cluster
$ kubectl exec -it netshoot-794598cdc4-x67z9 -- sh
/ # curl 10.8.7.222
...
...

# Spent 1 min trying to figure out why app was not working. Then realized that I should have done curl 10.8.7.222:8080
```

If it runs on port 80, is exposed on port 80, then it makes curling it for quick demos even easier and more automatic. Also, using consistent port in the app and also for pod exposure makes it simpler to understand because you're not trying to map different ports in your head. This is a demo-focused app so all of these things help to make the demos as simple and clean as they can be.

Reply:
Also awesome work on making the one-liner work! That is so easy and automatic that it will make this great for tutorials.

Reply:
chatted about this in person already, but the issue with port 80 is around permissions. unless the app runs as root within the container, you'll get a permission denied error: ```PermissionError: [Errno 13] Permission denied```